File: GraphicsTechniques/TextRenderingMsdf/ref/msdfText.ts
import { mat4, Mat4 } from 'wgpu-matrix';
import { MsdfFont } from "./MsdfFont";
import { MsdfTextMeasurements } from "./MsdfTextMeasurements";


// The kerning map stores a spare map of character ID pairs with an associated
// X offset that should be applied to the character spacing when the second
// character ID is rendered after the first.
export type KerningMap = Map<number, Map<number, number>>;

export class MsdfText {
  private bufferArray = new Float32Array(24);
  private bufferArrayDirty = true;

  constructor(
    public device: GPUDevice,
    private renderBundle: GPURenderBundle,
    public measurements: MsdfTextMeasurements,
    public font: MsdfFont,
    public textBuffer: GPUBuffer
  ) {
    mat4.identity(this.bufferArray);
    this.setColor(1, 1, 1, 1);
    this.setPixelScale(1 / 512);
    this.bufferArrayDirty = true;
  }

  getRenderBundle() {
    if (this.bufferArrayDirty) {
      this.bufferArrayDirty = false;
      this.device.queue.writeBuffer(
        this.textBuffer,
        0,
        this.bufferArray,
        0,
        this.bufferArray.length
      );
    }
    return this.renderBundle;
  }

  setTransform(matrix: Mat4) {
    mat4.copy(matrix, this.bufferArray);
    this.bufferArrayDirty = true;
  }

  setColor(r: number, g: number, b: number, a: number = 1.0) {
    this.bufferArray[16] = r;
    this.bufferArray[17] = g;
    this.bufferArray[18] = b;
    this.bufferArray[19] = a;
    this.bufferArrayDirty = true;
  }

  setPixelScale(pixelScale: number) {
    this.bufferArray[20] = pixelScale;
    this.bufferArrayDirty = true;
  }
}

File: GraphicsTechniques/TextRenderingMsdf/MsdfText.cs
using System.Numerics;
using System.Runtime.InteropServices;
using WebGpuSharp;
using GPUBuffer = WebGpuSharp.Buffer;

sealed class MsdfText
{
    private readonly Queue _queue;
    private FormattedTextStorage _bufferData;
    private bool _bufferDataDirty = true;
    private RenderBundle _renderBundle;

    public MsdfTextMeasurements Measurements { get; }
    public MsdfFont Font { get; }
    public GPUBuffer TextBuffer { get; }

    public MsdfText(
        Queue queue,
        RenderBundle renderBundle,
        MsdfTextMeasurements measurements,
        MsdfFont font,
        GPUBuffer textBuffer)
    {
        _queue = queue;
        _renderBundle = renderBundle;
        Measurements = measurements;
        Font = font;
        TextBuffer = textBuffer;

        SetTransform(Matrix4x4.Identity);
        SetColor(1, 1, 1, 1);
        SetPixelScale(1f / 512f);
        _bufferDataDirty = true;
    }

    public RenderBundle GetRenderBundle()
    {
        if (_bufferDataDirty)
        {
            _bufferDataDirty = false;
            _queue.WriteBuffer(TextBuffer, 0, _bufferData);
        }
        return _renderBundle;
    }

    public void SetTransform(in Matrix4x4 matrix)
    {
        _bufferData.Transform = matrix;
        _bufferDataDirty = true;
    }

    public void SetColor(float r, float g, float b, float a = 1.0f)
    {
        _bufferData.Color = new Vector4(r, g, b, a);
        _bufferDataDirty = true;
    }

    public void SetPixelScale(float pixelScale)
    {
        _bufferData.Scale = pixelScale;
        _bufferDataDirty = true;
    }
}
